README.md

<p align="center">
  <a href="https://app.logdna.com">
    <img height="95" width="201" src="https://raw.githubusercontent.com/logdna/artwork/master/logo%2Bruby.png">
  </a>
  <p align="center">Ruby gem for logging to <a href="https://app.logdna.com">LogDNA</a></p>
</p>

---

* **[Installation](#installation)**
* **[Quick Setup](#quick-setup)**
* **[API](#api)**
* **[Contributing](#contributing)**
* **[License](#license)**

# Installation

Add this line to your application's Gemfile:

```ruby
gem 'logdna'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install logdna



# Quick Setup

After installation, call

    logger = Logdna::Ruby.new(your_api_key, options)
    => #<Logdna::Ruby:0x00000000000000>

to set up the logger.

Options are optional variables that may contain hostname, app name, mac address, ip address, log level specified.

    options = {
        :hostname => myHostName,
        :ip =>  myIpAddress,
        :mac => myMacAddress,
        :app => myAppName,
        :level => "INFO",    # LOG_LEVELS = ['TRACE', 'DEBUG', 'INFO', 'WARN', 'ERROR', 'FATAL'] or your customized log level (custom levels for Rails have to be sent with a log message)
        :env => "PRODUCTION",
        :meta => {:once => {:first => "nested1", :another => "nested2"}},
        :endpoint => "https://fqdn/logs/ingest"
    }

To send logs, use "log" method. Default log level is "INFO"

    logger.log('This is my first log')
    => "Saved"  # Saved to buffer. Ready to be flushed automatically

Optionally you can use a block to do so

    logger.log { 'This is my second log' }
    => "Saved"

Log a message with particular metadata, level, appname, environment (one-time)

    logger.log('This is warn message', {:meta => {:meta => "data"}, :level => "WARN", :app => "awesome", :env => "DEVELOPMENT"})

Log a message with lasting metadata, level, appname, environment (lasting)

    logger.meta = {:once => {:first => "nested1", :another => "nested2"}}
    logger.level = 'FATAL'  or  logger.level = Logger::FATAL
    logger.app = 'NEW APP NAME'
    logger.env = 'PRODUCTION'
    logger.log('This messages and messages afterwards all have the above values')

Clear current metadata, level, appname, environment

    logger.clear

Check current log level:
    logger.info? => true
    logger.warn? => false

Log a message with a particular level easily

    logger.warn('This is a warning message')
    logger.fatal('This is a fatal message')
    logger.debug { 'This is a debug message' }


Hostname and app name cannot be more than 80 characters.

### Rails Setup
In your `config/environments/environment.rb`:

  ```
  Rails.application.configure do
    config.logger = Logdna::Ruby.new(your_api_key, options)
  end
  ```

# Important Notes

1. This logger assumes that you pass in json formatted data
2. This logger is a singleton (do not create mutiple instances of the logger) even though the singleton structure is not strongly enforced.

# API

## Logdna::Ruby.new(ingestion_key, options = {})

Instantiates a new instance of the class it is called on. ingestion_key is required.

| Options | Default |
|---------|---------|
|{ :hostname => Host name } | Device's default hostname |
|{ :mac => MAC address } | Nil |
|{ :ip => IP address } | Nil |
|{ :app => App name } | 'default' |
|{ :level => Log level } | 'INFO' |
|{ :env => STAGING, PRODUCTION .. etc} | Nil |
|{ :meta => metadata} | Nil |
|{ :endpoint => LogDNA Ingestion URI | 'https://logs.logdna.com/logs/ingest' |
|{ :flush_interval => Limit to trigger a flush in seconds } | 0.25 seconds |
|{ :flush_size => Limit to trigger a flush in bytes } | 2097152 bytes = 2 MiB |
|{ :request_size => Upper limit of request in bytes } | 2097152 bytes = 2 MiB |
|{ :retry_timeout => Base timeout for retries in seconds } | 0.25 seconds |
|{ :retry_max_attempts => Maximum number of retries per request } | 3 attempts |
|{ :retry_max_jitter => Maximum amount of jitter to add to each retry request in seconds } | 0.25 seconds |

Different log level displays log messages in different colors as well.
- ![TRACE DEBUG INFO Colors](https://placehold.it/15/515151/000000?text=+)   "Trace"  "Debug"  "Info"
- ![WARN Color](https://placehold.it/15/ec9563/000000?text=+)   "Warn"
- ![ERROR Fatal Colors](https://placehold.it/15/e37e7d/000000?text=+)   "Error"  "Fatal"



# Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/logdna/ruby.



# License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [v1.5.0] - 2021-01-29
### Added
- `request_size` option to control the upper limit of a request.
- `retry_timeout` option to control the base timeout for retries.
- `retry_max_attempts` option to control maximum number of retries.
- `retry_max_jitter` option to control the maximum amount of jitter (random time added) for each tried request.

### Changed
- Set the minimum required ruby version to 2.5.0.

### Fixed
- The client sometimes blocking the main thread during requests.

## [v1.4.2] - 2020-03-13
### Fixed
- Schedule flusher blocks main thread.

## [v1.4.1] - 2020-03-19
### Fixed
- Missing level method causing invalid log level error.

## [v1.4.0] - 2020-01-30
### Added
- Everything.

[Unreleased]: https://github.com/logdna/ruby/compare/v1.5.0...master
[v1.5.0]: https://github.com/logdna/ruby/compare/v1.4.2...v1.5.0
[v1.4.2]: https://github.com/logdna/ruby/compare/v1.4.1...v1.4.2
[v1.4.1]: https://github.com/logdna/ruby/compare/v1.4.0...v1.4.1
[v1.4.0]: https://github.com/logdna/ruby/releases/tag/v1.4.0


test.rb
require 'require_all'
require_all 'lib'

options = { hostname: "new_ruby", meta: { :once => { :first => "nested1", :another => "nested2" } } }

logger1 = Logdna::Ruby.new('Your API Key', options)

logger1.log('**************** This is the start of test ****************')
logger1.env = 'STAGING'
logger1.app = 'HELLO'
logger1.warn('Warn message with Staging and Hello')
logger1.clear
logger1.log('Is everything back to normal?')

logger1.log('Testing env app name change using log')
logger1.env = 'PRODUCTION'
logger1.app = 'CHANGED'
logger1.log('This should be stage = PRODUCTION and appname = CHANGED')
logger1.log('Testing env app name change using other messages')

logger1.error('This is error message with env = DEVELOPMENT and appname = NIHAO', { :env => 'DEVELOPMENT', :app => 'NIHAO' })
logger1.log('Should not stay as DEVELOPMENT and NIHAO')
logger1.env = 'DEVELOPMENT'
logger1.app = 'NIHAO'
logger1.log('Now should be DEVELOPMENT and NIHAO')
logger1.log('Logging metadata in trace level', { :meta => { :once => { :first => "nested1", :another => "nested2" } }, :level => "TRACE" })

logger1.level = Logger::DEBUG
logger1.log('This is debug message')
logger1.add('this should not be supported')
logger1.fatal('Does this continue as fatal?')
logger1.log('This should be debug')
logger1.level = Logger::WARN
logger1.log('**************** This is the end of test ****************')

=begin
logger1.level = Logger::WARN
logger1.log('This should be warn')
logger1.trace('This should be trace')
logger1.log('Again warn level')

logger1.log('Warn level log1')
logger1.info('Info level log1')
logger1.level = Logger::DEBUG
logger1.log('DEBUG log1')

logger1.app = 'NEW APP NAME'
logger1.env = 'Staging'
logger1.level = 'INFO'

logger1.level = 'INFO'
logger1.level == Logger::INFO

logger1.log('are changes all updated')
=end
sleep 3


tests.rb
# frozen_string_literal: true

require "minitest/autorun"

require_relative "lib/logdna.rb"
require_relative "lib/logdna/client.rb"
require_relative "test_server.rb"

class TestLogDNARuby < Minitest::Test
  LOG_LINE = "log line"

  def get_options(port)
    {
      hostname: "rubyTestHost",
      ip: "75.10.4.81",
      mac: "00:00:00:a1:2b:cc",
      app: "rubyApplication",
      level: "INFO",
      env: "PRODUCTION",
      endpoint: "http://localhost:#{port}",
      flush_interval: 1,
      flush_size: 5,
      retry_timeout: 1
    }
  end

  def log_level_test(level, port, expected_level)
    options = get_options(port)
    logdna_thread = Thread.start do
      logger = Logdna::Ruby.new("pp", options)
      logger.send(level, LOG_LINE)
    end

    server_thread = Thread.start do
      server_generator = TestServer.new
      server = server_generator.start_server(port)
      recieved_data = server_generator.accept_logs_and_respond(server, "HTTP/1.1 200 OK")

      assert_equal(recieved_data[:ls][0][:line], LOG_LINE)
      assert_equal(recieved_data[:ls][0][:app], options[:app])
      assert_equal(recieved_data[:ls][0][:level], expected_level)
      assert_equal(recieved_data[:ls][0][:env], options[:env])
    end

    logdna_thread.join
    server_thread.join
  end

  # Should retry to connect and preserve the failed line
  def retry_test(level, port, _expected_level)
    second_line = " second line"
    options = get_options(port)

    logger = Logdna::Ruby.new("pp", options)
    logger.send(level, LOG_LINE)

    server_thread = Thread.start do
      server_generator = TestServer.new
      server = server_generator.start_server(port)
      server_generator.accept_logs_and_respond(server, "HTTP/1.1 404 Not Found")
      # make a second request
      logger.send(level, second_line)
      recieved_data = server_generator.accept_logs_and_respond(server, "HTTP/1.1 202 Not Found")

      # should contain lines from both requests
      assert_equal(recieved_data[:ls].size, 2)
      # The order of recieved lines is unpredictable.
      assert_includes([recieved_data[:ls][0][:line], recieved_data[:ls][1][:line]], LOG_LINE)
      assert_includes([recieved_data[:ls][0][:line], recieved_data[:ls][1][:line]], second_line)
    end

    server_thread.join
  end

  def test_all
    log_level_test("warn", 2000, "WARN")
    log_level_test("info", 2001, "INFO")
    log_level_test("fatal", 2002, "FATAL")
    log_level_test("debug", 2003, "DEBUG")
    retry_test("fatal", 2004, "FATAL")
  end
end


test/tests.rb
# frozen_string_literal: true

require "minitest/autorun"

require_relative "../lib/logdna"
require_relative "../lib/logdna/client"
require_relative "test_server"

class TestLogDNARuby < Minitest::Test
  LOG_LINE = "log line"

  def get_options(port)
    {
      hostname: "rubyTestHost",
      ip: "75.10.4.81",
      mac: "00:00:00:a1:2b:cc",
      app: "rubyApplication",
      level: "INFO",
      env: "PRODUCTION",
      endpoint: "http://localhost:#{port}",
      flush_interval: 1,
      flush_size: 5,
      retry_timeout: 1
    }
  end

  def log_level_test(level, port, expected_level)
    options = get_options(port)
    logdna_thread = Thread.start do
      logger = Logdna::Ruby.new("pp", options)
      logger.send(level, LOG_LINE)
    end

    server_thread = Thread.start do
      server_generator = TestServer.new
      server = server_generator.start_server(port)
      recieved_data = server_generator.accept_logs_and_respond(server, "HTTP/1.1 200 OK")

      assert_equal(recieved_data[:ls][0][:line], LOG_LINE)
      assert_equal(recieved_data[:ls][0][:app], options[:app])
      assert_equal(recieved_data[:ls][0][:level], expected_level)
      assert_equal(recieved_data[:ls][0][:env], options[:env])
    end

    assert logdna_thread.join(10), "Logdna thread timeout for port #{port}"
    assert server_thread.join(10), "Server thread timeout for port #{port}"
  end

  # Should retry to connect and preserve the failed line
  def retry_test(level, port, _expected_level)
    second_line = " second line"
    options = get_options(port)

    logger = Logdna::Ruby.new("pp", options)
    logger.send(level, LOG_LINE)

    server_thread = Thread.start do
      server_generator = TestServer.new
      server = server_generator.start_server(port)
      server_generator.accept_logs_and_respond(server, "HTTP/1.1 408 Request Timeout")
      # make a second request
      logger.send(level, second_line)

      # Both lines will come back in separate requests
      recieved_data1 = server_generator.accept_logs_and_respond(server, "HTTP/1.1 200 OK")
      recieved_data2 = server_generator.accept_logs_and_respond(server, "HTTP/1.1 200 OK")

      # The order of recieved lines is unpredictable.
      assert_includes([recieved_data1[:ls][0][:line], recieved_data2[:ls][0][:line]], LOG_LINE)
      assert_includes([recieved_data1[:ls][0][:line], recieved_data2[:ls][0][:line]], second_line)
    end

    assert server_thread.join(10), "Server thread timeout for retry test"
  end

  def test_all
    log_level_test("warn", 2000, "WARN")
    log_level_test("info", 2001, "INFO")
    log_level_test("fatal", 2002, "FATAL")
    log_level_test("debug", 2003, "DEBUG")
    retry_test("fatal", 2004, "FATAL")
  end
end


lib/logdna.rb
# frozen_string_literal: true

require "logger"
require "socket"
require "uri"
require_relative "logdna/client"
require_relative "logdna/resources"
require_relative "logdna/version"

module Logdna
  class ValidURLRequired < ArgumentError; end

  class MaxLengthExceeded < ArgumentError; end

  class Ruby < ::Logger
    # uncomment line below and line 3 to enforce singleton
    # include Singleton
    Logger::TRACE = 5
    attr_accessor :app, :env, :meta

    def initialize(key, opts = {})
      super(nil, nil, nil)
      @app = opts[:app] || "default"
      @log_level = opts[:level] || "INFO"
      @env = opts[:env]
      @meta = opts[:meta]
      @internal_logger = Logger.new($stdout)
      @internal_logger.level = Logger::DEBUG
      endpoint = opts[:endpoint] || Resources::ENDPOINT
      hostname = opts[:hostname] || Socket.gethostname

      if hostname.size > Resources::MAX_INPUT_LENGTH || @app.size > Resources::MAX_INPUT_LENGTH
        @internal_logger.debug("Hostname or Appname is over #{Resources::MAX_INPUT_LENGTH} characters")
        return
      end

      ip =  opts.key?(:ip) ? "&ip=#{opts[:ip]}" : ""
      mac = opts.key?(:mac) ? "&mac=#{opts[:mac]}" : ""
      url = "#{endpoint}?hostname=#{hostname}#{mac}#{ip}"
      uri = URI(url)

      request = Net::HTTP::Post.new(uri.request_uri, "Content-Type" => "application/json")
      request.basic_auth("username", key)
      request[:'user-agent'] = opts[:'user-agent'] || "ruby/#{LogDNA::VERSION}"
      @client = Logdna::Client.new(request, uri, opts)
    end

    def default_opts
      {
        app: @app,
        level: @log_level,
        env: @env,
        meta: @meta,
      }
    end

    def level
      @log_level
    end

    def level=(value)
      if value.is_a? Numeric
        @log_level = Resources::LOG_LEVELS[value]
        return
      end

      @log_level = value
    end

    def log(message = nil, opts = {})
      if message.nil? && block_given?
        message = yield
      end
      if message.nil?
        @internal_logger.debug("provide either a message or block")
        return
      end
      message = message.to_s.encode("UTF-8")
      @client.write_to_buffer(message, default_opts.merge(opts).merge(
                                         timestamp: (Time.now.to_f * 1000).to_i
                                       ))
    end

    Resources::LOG_LEVELS.each do |lvl|
      name = lvl.downcase

      define_method name do |msg = nil, opts = {}, &block|
        self.log(msg, opts.merge(
                        level: lvl
                      ), &block)
      end

      define_method "#{name}?" do
        return Resources::LOG_LEVELS[self.level] == lvl if level.is_a? Numeric

        self.level == lvl
      end
    end

    def clear
      @app = "default"
      @log_level = "INFO"
      @env = nil
      @meta = nil
    end

    def <<(msg = nil, opts = {})
      log(msg, opts.merge(
                 level: ""
               ))
    end

    def add(*_arg)
      @internal_logger.debug("add not supported in LogDNA logger")
      false
    end

    def unknown(msg = nil, opts = {})
      log(msg, opts.merge(
                 level: "UNKNOWN"
               ))
    end

    def datetime_format(*_arg)
      @internal_logger.debug("datetime_format not supported in LogDNA logger")
      false
    end

    def close
      @client&.exitout
    end
  end
end


test-server.rb
require 'socket'
class TestServer
  attr_accessor :a
  def startServer(port)
      server = TCPServer.new(port)
      puts server
      data = ''
      
      Thread.start(server.accept) { |client|    
      headers = {}
        while line = client.gets.split(' ', 2)
            break if line[0] == ""
            headers[line[0].chop] = line[1].strip
        end
        data = client.read(headers["Content-Length"].to_i)
        client.puts("HTTP/1.1 200 OK")
        client.close
      }.join
   
    return eval(data)
  end
  
  def startServerWithNotFound(port)
    server = TCPServer.new(port)
    
    count = 0 
    data = ''
    loop do
      count = count + 1
      Thread.start(server.accept) { |client|    
      headers = {}
        while line = client.gets.split(' ', 2)
            break if line[0] == ""
            headers[line[0].chop] = line[1].strip
        end
        data =+ client.read(headers["Content-Length"].to_i)
        if (count < 2) 
          client.puts("HTTP/1.1 404 Not Found")
        else
          client.puts("HTTP/1.1 200 OK")
        end
        client.close
      }.join  
      break if count == 2
    end
  return eval(data)
end
end




lib/logdna/client.rb
# frozen_string_literal: true

require "etc"
require "net/http"
require "socket"
require "json"
require "concurrent"
require "date"
require "securerandom"

module Logdna
  Message = Struct.new(:source, :running_size)

  class Client
    def initialize(request, uri, opts)
      @uri = uri

      # NOTE: buffer is in memory
      @buffer = []

      @lock = Mutex.new

      @flush_interval = opts[:flush_interval] || Resources::FLUSH_INTERVAL
      @flush_size = opts[:flush_size] || Resources::FLUSH_SIZE

      @request = request
      @request_size = opts[:request_size] || Resources::REQUEST_SIZE

      @retry_timeout = opts[:retry_timeout] || Resources::RETRY_TIMEOUT
      @retry_max_jitter = opts[:retry_max_jitter] || Resources::RETRY_MAX_JITTER
      @retry_max_attempts = opts[:retry_max_attempts] || Resources::RETRY_MAX_ATTEMPTS

      @internal_logger = Logger.new($stdout)
      @internal_logger.level = Logger::DEBUG

      @work_thread_pool = Concurrent::FixedThreadPool.new(Etc.nprocessors)
      # TODO: Expose an option to configure the maximum concurrent requests
      # Requires the instance-global request to be resolved first
      @request_thread_pool = Concurrent::FixedThreadPool.new(Resources::MAX_CONCURRENT_REQUESTS)

      @scheduled_flush = nil
    end

    def schedule_flush
      if @scheduled_flush.nil? || @scheduled_flush.complete?
        @scheduled_flush = Concurrent::ScheduledTask.execute(@flush_interval) { flush }
      end
    end

    def unschedule_flush
      if !@scheduled_flush.nil?
        @scheduled_flush.cancel
        @scheduled_flush = nil
      end
    end

    def process_message(msg, opts = {})
      processed_message = {
        line: msg,
        app: opts[:app],
        level: opts[:level],
        env: opts[:env],
        meta: opts[:meta],
        timestamp: Time.now.to_i,
      }
      processed_message.delete(:meta) if processed_message[:meta].nil?
      processed_message
    end

    def write_to_buffer(msg, opts)
      Concurrent::Future.execute({ executor: @work_thread_pool }) { write_to_buffer_sync(msg, opts) }
    end

    def write_to_buffer_sync(msg, opts)
      processed_message = process_message(msg, opts)
      message_size = processed_message.to_s.bytesize

      running_size = @lock.synchronize do
        running_size = message_size
        if @buffer.any?
          running_size += @buffer[-1].running_size
        end
        @buffer.push(Message.new(processed_message, running_size))

        running_size
      end

      if running_size >= @flush_size
        unschedule_flush
        flush_sync
      else
        schedule_flush
      end
    end

    ##
    # Flushes all logs to LogDNA asynchronously
    def flush(options = {})
      Concurrent::Future.execute({ executor: @work_thread_pool }) { flush_sync(options) }
    end

    ##
    # Flushes all logs to LogDNA synchronously
    def flush_sync(options = {})
      slices = @lock.synchronize do
        # Slice the buffer into chunks that try to be no larger than @request_size. Slice points are found with
        # a binary search thanks to the structure of @buffer. We are working backwards because it's cheaper to
        # remove from the tail of an array instead of the head
        slices = []
        until @buffer.empty?
          search_size = @buffer[-1].running_size - @request_size
          if search_size.negative?
            search_size = 0
          end

          slice_index = @buffer.bsearch_index { |message| message.running_size >= search_size }
          slices.push(@buffer.pop(@buffer.length - slice_index).map(&:source))
        end
        slices
      end

      # Remember the chunks are in reverse order, this un-reverses them
      slices.reverse_each do |slice|
        if options[:block_on_requests]
          try_request(slice)
        else
          Concurrent::Future.execute({ executor: @request_thread_pool }) { try_request(slice) }
        end
      end
    end

    def try_request(slice)
      body = {
        e: "ls",
        ls: slice
      }.to_json

      flush_id = "#{SecureRandom.uuid} [#{slice.length} lines]"
      error_header = "Flush {#{flush_id}} failed."
      tries = 0
      loop do
        tries += 1

        if tries > @retry_max_attempts
          @internal_logger.debug("Flush {#{flush_id}} exceeded 3 tries. Discarding flush buffer")
          break
        end

        if send_request(body, error_header)
          break
        end

        sleep(@retry_timeout * (1 << (tries - 1)) + rand(@retry_max_jitter))
      end
    end

    def send_request(body, error_header)
      # TODO: Remove instance-global request object
      @request.body = body
      begin
        response = Net::HTTP.start(
          @uri.hostname,
          @uri.port,
          use_ssl: @uri.scheme == "https"
        ) do |http|
          http.request(@request)
        end

        code = response.code.to_i
        if [401, 403].include?(code)
          @internal_logger.debug("#{error_header} Please provide a valid ingestion key. Discarding flush buffer")
          return true
        elsif [408, 500, 504].include?(code)
          # These codes might indicate a temporary ingester issue
          @internal_logger.debug("#{error_header} The request failed #{response}. Retrying")
        elsif code == 200
          return true
        else
          @internal_logger.debug("#{error_header} The request failed #{response}. Discarding flush buffer")
          return true
        end
      rescue SocketError
        @internal_logger.debug("#{error_header} Network connectivity issue. Retrying")
      rescue Errno::ECONNREFUSED => e
        @internal_logger.debug("#{error_header} The server is down. #{e.message}. Retrying")
      rescue Timeout::Error => e
        @internal_logger.debug("#{error_header} Timeout error occurred. #{e.message}. Retrying")
      end

      false
    end

    def exitout
      unschedule_flush
      @work_thread_pool.shutdown
      if !@work_thread_pool.wait_for_termination(1)
        @internal_logger.warn("Work thread pool unable to shutdown gracefully. Logs potentially dropped")
      end
      @request_thread_pool.shutdown
      if !@request_thread_pool.wait_for_termination(5)
        @internal_logger.warn("Request thread pool unable to shutdown gracefully. Logs potentially dropped")
      end

      if @buffer.any?
        @internal_logger.debug("Exiting LogDNA logger: Logging remaining messages")
        flush_sync({ block_on_requests: true })
        @internal_logger.debug("Finished flushing logs to LogDNA")
      end
    end
  end
end
